components/feature/media/src/main/java/mozilla/components/feature/media/middleware/LastMediaAccessMiddleware.kt

```diff
@@ -4,34 +4,196 @@
+                mediaTabsStatus.clear()
+            }
+            // If a media tab was closed but then followed by "Undo" or when restored after the app is restarted
+            // we need to keep tracking it's media status.
+            is TabListAction.RestoreAction -> {
+                action.tabs
+                    .filter { it.lastMediaAccess > MEDIA_NOT_PLAYING_TIMESTAMP }
+                    .forEach {
+                        rememberTabStatusForMediaPlaying(it, it.lastMediaAccess)
+                    }
+            }
+            else -> {
+                // no-op
+            }
+        }
+    }
+
+    @VisibleForTesting
+    internal fun rememberTabStatusForMediaPlaying(mediaTab: TabSessionState, mediaStartedPlayingTimeStamp: Long) {
+        mediaTabsStatus[mediaTab.id] = MediaTabStatus(
+            urlWhenMediaStarted = mediaTab.content.url,
+            lastMediaAccess = mediaStartedPlayingTimeStamp
+        )
+    }
+
+    @VisibleForTesting
+    internal fun maybeUpdateTabStatusForUrlUpdated(tab: TabSessionState, updatedUrl: String) {
+        mediaTabsStatus.maybeUpdate(tab.id) {
+            it.copy(updatedTabUrl = updatedUrl)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun maybeUpdateTabStatusForMediaEnded(tab: TabSessionState) {
+        mediaTabsStatus.maybeUpdate(tab.id) {
+            it.copy(lastMediaAccess = MEDIA_NOT_PLAYING_TIMESTAMP)
+        }
+    }
+
+    @VisibleForTesting
+    internal fun maybeDispatchPlaybackEnded(
+        tab: TabSessionState,
+        context: MiddlewareContext<BrowserState, BrowserAction>
+    ) {
+        mediaTabsStatus[tab.id]?.let {
+            if (it.hasMediaEnded()) {
+                context.dispatch(
+                    LastAccessAction.UpdateLastMediaAccessAction(tab.id, MEDIA_NOT_PLAYING_TIMESTAMP)
+                )
+            }
         }
     }
 }
+
+/**
+ * Media tab details based on which we know playback ended.
+ *
+ * @property urlWhenMediaStarted URL of the page on which media started playing.
+ * @property updatedTabUrl current URL of the current tab.
+ * If this is different than [urlWhenMediaStarted] then the user probably navigated to another pag
+ * and media stopped playing.
+ * Users could also navigate in the same HTML document to different URLs but with media continuing to play.
+ * @property lastMediaAccess MEDIA_NOT_PLAYING_TIMESTAMP if media was playing in this tab but then stopped
+ * or a value bigger than MEDIA_NOT_PLAYING_TIMESTAMP signifying the timestamp of media starting to play.
+ */
+@VisibleForTesting
+internal data class MediaTabStatus(
+    val urlWhenMediaStarted: String,
+    val updatedTabUrl: String = urlWhenMediaStarted,
+    val lastMediaAccess: Long = MEDIA_NOT_PLAYING_TIMESTAMP
+)
+
+/**
+ * Filter all tabs ignoring custom ones and execute [action] on the normal/private tab with [tabId] if it exists.
+ */
+@VisibleForTesting
+internal fun MiddlewareContext<BrowserState, BrowserAction>.ifNormalTab(
```

Comment:
Something isn't right here. The method name seems to suggest that this only runs for `normal` tabs but the implementation doesn't filter and also runs it for private tabs.

Did you mean to use `state.normalTabs` or filter here? We should also move this method to `Selectors.kt` to our other extension functions. I think it will be useful in the future.

Reply:
I wanted to exclude custom tabs.
So we'd remain with the other "normal" ones. Which are normal and private tabs. But I don't know how to better refer to them as a group and make the distinction from custom tabs clearer.
Maybe just call them by name and in the event this kind of method will be needed in the new implementation it could be named `ifNormalOrPrivateTab`?

Reply:
@Mugurell yes, exactly, we use `normal` to indicate non-private. That was the confusing part :).